File: src/languages/sundanese/sundanese.ts
/**
 * Native Sundanese / Basa Sunda (su) text phonemizer — canonical IPA. Austronesian (West
 * Java), modern LATIN orthography (+ the revived Aksara Sunda abugida, transliterated to Latin by sundaAksara.ts — identical IPA). Shallow and near-phonemic (the id/jv pattern), so a flat left-to-right scan:
 * digraphs (the central vowel ⟨eu⟩→[ɨ], ng→[ŋ], ny→[ɲ]) then single letters, ⟨e⟩→schwa [ə] / ⟨é⟩→[e], c→[t͡ʃ],
 * j→[d͡ʒ]. Glottal stop is inserted at a word-initial vowel (awi→ʔawi) and in a same-vowel hiatus (naam→naʔam).
 * Penultimate (weak) stress, skipping a schwa nucleus. Validated against kaikki su.
 */
import type { Phonemizer } from "../../registry.ts";
import { assembleClauses } from "../../core/clauses.ts";
import { hostWordRun, makeNativiser } from "../../core/hostWord.ts";
import { loadManifest } from "../../core/loadManifest.ts";
import { aksaraToLatin, isAksaraSunda, normalizeSundaDigits } from "./sundaAksara.ts";
import { normalizeSundanese } from "./normalize.ts";

interface NumbersDef {
    units: string[];
    belas: string;
    puluh: string;
    ratus: string;
    rebu: string;
    yuta: string;
    seprefix: string;
}
interface SundaneseDef {
    digraphs: Record<string, string>;
    vowels: Record<string, string>;
    consonants: Record<string, string>;
    glottal: string;
    numbers: NumbersDef;
    clausePunctuation: Record<string, string>;
}
const DEF = loadManifest<SundaneseDef>(import.meta.url, "sundanese.jsonc");
const CLAUSE_MARK = DEF.clausePunctuation;
const NUM = DEF.numbers;
const VOWEL_PH = "aeiouəɨ";
const isVowelPh = (s: string): boolean => VOWEL_PH.includes(s);

/** One Sundanese word → canonical IPA. Accepts BOTH scripts: the Latin orthography and Aksara Sunda (ᮃᮊ᮪ᮞᮛ) —
 *  Aksara Sunda is transliterated to Latin first (identical IPA), then the shared g2p + glottal + stress runs. */
export function phonemizeWord(word: string): string {
    const latin = isAksaraSunda(word) ? aksaraToLatin(word) : word;
    const s = [...latin.toLowerCase().normalize("NFC")];
    const segs: string[] = [];
    for (let i = 0; i < s.length; ) {
        const two = s[i]! + (s[i + 1] ?? "");
        if (DEF.digraphs[two] !== undefined) {
            segs.push(DEF.digraphs[two]!);
            i += 2;
            continue;
        }
        const c = s[i]!;
        if (DEF.vowels[c] !== undefined) segs.push(DEF.vowels[c]!);
        else if (DEF.consonants[c] !== undefined) segs.push(DEF.consonants[c]!);
        // else: unknown char → skip
        i++;
    }
    // Glottal stop: word-initial vowel (ʔawi), and between two IDENTICAL adjacent vowels (naam→naʔam).
    if (segs.length && isVowelPh(segs[0]!)) segs.unshift(DEF.glottal);
    for (let i = segs.length - 1; i > 0; i--) {
        if (isVowelPh(segs[i]!) && segs[i] === segs[i - 1]) segs.splice(i, 0, DEF.glottal);
    }
    // Penultimate (weak) stress on the penult vowel nucleus — but a schwa penult can't bear stress, so it shifts
    // to the final vowel (the Indonesian/Malay pattern: hese→hesˈe, not hˈəse).
    const vidx = segs.map((ph, idx) => (isVowelPh(ph) ? idx : -1)).filter((x) => x >= 0);
    if (vidx.length >= 2) {
        const penult = vidx[vidx.length - 2]!;
        const at = segs[penult] === "ə" ? vidx[vidx.length - 1]! : penult;
        segs.splice(at, 0, "ˈ");
    }
    return segs.join("").normalize("NFC");
}

// Austronesian decimal composition (Sundanese): units + -belas teens + -puluh tens + ratus/rebu/yuta; a leading
// "1" of a magnitude is the sa- prefix (sapuluh, saratus). Word forms are read back through the g2p.
function toWords(n: number): string {
    if (n < 10) return NUM.units[n]!;
    if (n < 20) return n === 10 ? NUM.seprefix + NUM.puluh : n === 11 ? NUM.seprefix + NUM.belas : NUM.units[n - 10]! + " " + NUM.belas;
    if (n < 100) {
        const t = Math.floor(n / 10),
            u = n % 10;
        return NUM.units[t]! + " " + NUM.puluh + (u ? " " + NUM.units[u]! : "");
    }
    if (n < 1000) {
        const h = Math.floor(n / 100),
            r = n % 100;
        return (h === 1 ? NUM.seprefix + NUM.ratus : NUM.units[h]! + " " + NUM.ratus) + (r ? " " + toWords(r) : "");
    }
    if (n < 1_000_000) {
        const th = Math.floor(n / 1000),
            r = n % 1000;
        return (th === 1 ? NUM.seprefix + NUM.rebu : toWords(th) + " " + NUM.rebu) + (r ? " " + toWords(r) : "");
    }
    const m = Math.floor(n / 1_000_000),
        r = n % 1_000_000;
    return (m === 1 ? NUM.seprefix + NUM.yuta : toWords(m) + " " + NUM.yuta) + (r ? " " + toWords(r) : "");
}
function number(digits: string): string {
    const n = Number(digits);
    // ⚠ ABOVE 2^53 THE RAW ASCII DIGITS USED TO LEAK STRAIGHT INTO THE IPA. `isSafeInteger` is right to
    // refuse to COMPOSE — the float has already lost the low digits, so the numeral would be confidently
    // wrong — but the refusal returned the digit string, which no g2p in this fleet reads. Read it out
    // digit-at-a-time instead, THROUGH THE SAME COMPOSER: a one-digit number is a call this engine already
    // answers, so the fallback cannot invent a word. See core/numbers.ts `spellDigits` for the full
    // account and the cost — above 2^53 the reading is a digit string, not a quantity.
    const words = Number.isSafeInteger(n) ? toWords(n) : [...digits].map((d) => toWords(Number(d))).join(" ");
    return words.split(" ").map(phonemizeWord).join(" ");
}

// A word — Latin (incl. é) OR Aksara Sunda (letters + signs U+1B80–1BAF, 1BBA–1BBF; digits normalised to ASCII).
const TOKEN = new RegExp(`(${hostWordRun(["Latin", "Sundanese"])})|(\\d+)|([.?!,;:…])`, "giu");
/**
 * This language's OWN inventory — the TOKEN word class as it stood before the widening above, lifted verbatim, so
 * nothing about the orthography is invented here. A token this REJECTS carries a letter the language does not
 * use, i.e. a foreign name. See core/hostWord.ts: this is the INVENTORY question, and it is no longer also
 * deciding where the script boundary falls.
 */
const NATIVE_CLASS = "[a-zéÉ\\u{1B80}-\\u{1BAF}\\u{1BBA}-\\u{1BBF}]";
const nat = makeNativiser(NATIVE_CLASS, "iu");


class SundanesePhonemizer implements Phonemizer {
    text(input: string): string {
        // NORMALIZATION runs first — pure text→text, so everything it emits is then read by the ordinary word,
        // number and clause paths below. It must see the text BEFORE tokenization, because most of what it
        // repairs (a grouping `.`, a decimal `,`, a clock `:`) is a character `TOKEN` would otherwise hand to
        // `clausePunctuation` as a pause. ⚠ AFTER `normalizeSundaDigits`, so Aksara Sunda digits are already
        // ASCII and the number rules see one digit set rather than two.
        return assembleClauses(normalizeSundanese(normalizeSundaDigits(input)), TOKEN, (m, sink) => {
            if (m[1]) sink.emit(phonemizeWord(nat(m[1])));
            else if (m[2]) sink.emit(number(m[2]));
            else if (m[3]) {
                const mk = CLAUSE_MARK[m[3]];
                if (mk) sink.pause(mk);
            }
        });
    }
}

/** Build the Sundanese phonemizer. */
export function createSundanese(): Phonemizer {
    return new SundanesePhonemizer();
}

File: csharp/Vernacula.Phonemizer/Languages/Sundanese/Sundanese.cs
/**
 * Native Sundanese / Basa Sunda (su) text phonemizer — canonical IPA. Austronesian (West Java), modern LATIN
 * orthography plus the revived Aksara Sunda abugida (transliterated to Latin by SundaAksara — identical IPA).
 * Shallow and near-phonemic, so a flat left-to-right scan, glottal insertion, and penultimate stress.
 * Ported from src/languages/sundanese/sundanese.ts — see that file for the corpus evidence.
 */
using System.Text;
using Vernacula.Phonemizer.Core;

namespace Vernacula.Phonemizer.Languages.Sundanese;

public static class SundanesePhonemizer
{
    private static SundaneseDef DEF => Manifest.DEF;
    private static IReadOnlyDictionary<string, string> CLAUSE_MARK => DEF.ClausePunctuation;
    private static SundaneseNumbersDef NUM => DEF.Numbers;

    private const string VOWEL_PH = "aeiouəɨ";
    private static bool IsVowelPh(string s) => VOWEL_PH.Contains(s, StringComparison.Ordinal);

    /**
     * One Sundanese word → canonical IPA. Accepts BOTH scripts: the Latin orthography and Aksara Sunda,
     * which is transliterated to Latin first, then the shared g2p + glottal + stress runs.
     */
    public static string PhonemizeWord(string word)
    {
        var latin = SundaAksara.IsAksaraSunda(word) ? SundaAksara.AksaraToLatin(word) : word;
        var s = Js.CodePoints(Js.Normalize(Js.ToLowerCase(latin), NormalizationForm.FormC));
        var segs = new List<string>();
        for (var i = 0; i < s.Count;)
        {
            var two = s[i] + (i + 1 < s.Count ? s[i + 1] : "");
            if (DEF.Digraphs.TryGetValue(two, out var dg))
            {
                segs.Add(dg);
                i += 2;
                continue;
            }
            var c = s[i];
            if (DEF.Vowels.TryGetValue(c, out var v)) segs.Add(v);
            else if (DEF.Consonants.TryGetValue(c, out var k)) segs.Add(k);
            // else: unknown char → skip
            i++;
        }
        // Glottal stop: word-initial vowel (ʔawi), and between two IDENTICAL adjacent vowels (naam→naʔam).
        if (segs.Count > 0 && IsVowelPh(segs[0])) segs.Insert(0, DEF.Glottal);
        for (var i = segs.Count - 1; i > 0; i--)
            if (IsVowelPh(segs[i]) && segs[i] == segs[i - 1]) segs.Insert(i, DEF.Glottal);
        // Penultimate (weak) stress, shifted to the final vowel when the penult nucleus is a schwa.
        var vidx = new List<int>();
        for (var i = 0; i < segs.Count; i++) if (IsVowelPh(segs[i])) vidx.Add(i);
        if (vidx.Count >= 2)
        {
            var penult = vidx[^2];
            var at = segs[penult] == "ə" ? vidx[^1] : penult;
            segs.Insert(at, "ˈ");
        }
        return string.Concat(segs).Normalize(NormalizationForm.FormC);
    }

    /** Austronesian decimal composition: units + -belas teens + -puluh tens + ratus/rebu/yuta, sa- for a leading 1. */
    private static string ToWords(double n)
    {
        if (n < 10) return NUM.Units[(int)n];
        if (n < 20)
            return n == 10 ? NUM.Seprefix + NUM.Puluh
                : n == 11 ? NUM.Seprefix + NUM.Belas
                : NUM.Units[(int)n - 10] + " " + NUM.Belas;
        if (n < 100)
        {
            double t = Math.Floor(n / 10), u = n % 10;
            return NUM.Units[(int)t] + " " + NUM.Puluh + (u != 0 ? " " + NUM.Units[(int)u] : "");
        }
        if (n < 1000)
        {
            double h = Math.Floor(n / 100), r = n % 100;
            return (h == 1 ? NUM.Seprefix + NUM.Ratus : NUM.Units[(int)h] + " " + NUM.Ratus)
                + (r != 0 ? " " + ToWords(r) : "");
        }
        if (n < 1_000_000)
        {
            double th = Math.Floor(n / 1000), r = n % 1000;
            return (th == 1 ? NUM.Seprefix + NUM.Rebu : ToWords(th) + " " + NUM.Rebu)
                + (r != 0 ? " " + ToWords(r) : "");
        }
        double m = Math.Floor(n / 1_000_000), rm = n % 1_000_000;
        return (m == 1 ? NUM.Seprefix + NUM.Yuta : ToWords(m) + " " + NUM.Yuta)
            + (rm != 0 ? " " + ToWords(rm) : "");
    }

    private static string Number(string digits)
    {
        var n = Js.Number(digits);
        // Above 2^53 the float has already lost the low digits, so the numeral would be confidently wrong;
        // read it out digit-at-a-time THROUGH THE SAME COMPOSER. See core/numbers.ts `spellDigits`.
        var words = double.IsInteger(n) && Math.Abs(n) <= 9007199254740991d
            ? ToWords(n)
            : string.Join(" ", Js.CodePoints(digits).Select(d => ToWords(Js.Number(d))));
        return string.Join(" ", words.Split(' ').Select(PhonemizeWord));
    }

    // A word — Latin (incl. é) OR Aksara Sunda (letters + signs U+1B80–1BAF, 1BBA–1BBF).
    private static readonly JsRe TOKEN = JsRegex.Compile(
        $"({HostWord.HostWordRun(new[] { "Latin", "Sundanese" })})|(\\d+)|([.?!,;:…])", "giu");

    /** This language's OWN inventory — the INVENTORY question, not the script-boundary one. */
    private const string NATIVE_CLASS = "[a-zéÉ\\u{1B80}-\\u{1BAF}\\u{1BBA}-\\u{1BBF}]";
    private static readonly Func<string, string> Nat = HostWord.MakeNativiser(NATIVE_CLASS, "iu");

    private sealed class Engine : ILanguage
    {
        public string Text(string input)
        {
            // NORMALIZATION runs first, and AFTER NormalizeSundaDigits so the number rules see one digit set.
            return Clauses.AssembleClauses(
                Normalize.NormalizeSundanese(SundaAksara.NormalizeSundaDigits(input)), TOKEN, (m, sink) =>
                {
                    if (m.Groups[1].Success && m.Groups[1].Value.Length > 0) sink.Emit(PhonemizeWord(Nat(m.Groups[1].Value)));
                    else if (m.Groups[2].Success && m.Groups[2].Value.Length > 0) sink.Emit(Number(m.Groups[2].Value));
                    else if (m.Groups[3].Success && m.Groups[3].Value.Length > 0)
                    {
                        var mk = CLAUSE_MARK.GetValueOrDefault(m.Groups[3].Value);
                        if (!string.IsNullOrEmpty(mk)) sink.Pause(mk);
                    }
                });
        }
    }

    /** Build the Sundanese phonemizer. */
    public static ILanguage CreateSundanese() => new Engine();

    internal static void RegisterSelf() => Registry.Register("sundanese", () => CreateSundanese());
}
